mosaic_fb.c

```diff
@@ -1616,31 +1628,62 @@ void kalign_vt(struct data *my_data, struct pars *my_pars, struct matrices *my_m
 	/*Reset copy state*/
 	my_matrices->who_copy[target]=tmp_copy;
 
-	ofp = fopen(my_pars->alignment_file, "a");
+	ofp = fopen(my_pars->alignment_file, ALIGNMENT_FILE_APPEND);
 	fprintf(ofp,"\nTarget: %s\tLength: %i\tMLlk: %.3lf\n",my_data->seqs[target]->name, my_data->seqs[target]->length, my_matrices->llk);
 
+	const int printWidth = gDISPLAY_WIDTH;
+	const int coordWidth = 12;
+	char formatString[100];
+	snprintf(formatString, sizeof(formatString), "%%%ds [%%4d-%%4d]\t", printWidth);
+
 	/*First print target sequence*/
+	int tgtPosStart = -1;
+	int tgtPosEnd = -1;
 	cp++;
-	strncpy(tmp_name, my_data->seqs[target]->name, 15);
-	tmp_name[15]='\0';
-	fprintf(ofp,"%15s\t", tmp_name);
-	for (i=cp,pos_target=1;i<=2*my_matrices->maxl;i++) {
-		if (my_matrices->maxpath_state[i]==3) fprintf(ofp,"-");
+	strncpy(tmp_name, my_data->seqs[target]->name, printWidth);
+	tmp_name[printWidth] ='\0';
+	
+	// Create a place to hold the sequence alignment string:
+	unsigned int tmp_seq_pos = 0;
+	const unsigned int target_length = 1 + my_data->seqs[target]->length * 2;
+	char tmp_seq_string[target_length];
+	memset(tmp_seq_string, '\0', target_length);
+	
+	for (i=cp, pos_target=1; i <= 2*my_matrices->maxl; ++i) {
+		if (my_matrices->maxpath_state[i]==3) {
+			tmp_seq_string[tmp_seq_pos++] = '-';
+			tmp_seq_string[tmp_seq_pos] = '\0';
+		}
 		else {
-			fprintf(ofp,"%c",num2nuc(my_data->seqs[target]->seq[pos_target], my_data->type));
+			if ( tgtPosStart == -1 ) {
+				tgtPosStart = pos_target - 1;
+			}
+			tgtPosEnd = pos_target - 1;
+			
+			tmp_seq_string[tmp_seq_pos++] = num2nuc(my_data->seqs[target]->seq[pos_target], my_data->type);
+			tmp_seq_string[tmp_seq_pos] = '\0';
+			
 			pos_target++;
 		}
 	}
-	fprintf(ofp,"\n");
+	fprintf(ofp, formatString, tmp_name, tgtPosStart, tgtPosEnd);
+	fprintf(ofp, "%s", tmp_seq_string);
+	fprintf(ofp, "\n");
 	fflush(ofp);
 
 	/*Now do matching*/
-	for (i=1;i<=15;i++) fprintf(ofp," ");
+	for (i=1; i <= printWidth + coordWidth; ++i) {
+		fprintf(ofp," ");
+	}
 	fprintf(ofp,"\t");
-	for (i=cp, pos_target=1;i<=2*my_matrices->maxl;i++) {
+	for (i=cp, pos_target=1; i <= 2*my_matrices->maxl; ++i) {
```

Comment:
Does moving this increment from postfix to prefix not break anything?  Or does it fix something I wasn't aware was broken?

Reply:
It's a micro-optimization I got into the habit of when I was working on real-time systems.

Under the hood the compiler has to make a copy of the object being incremented in order to use its value before incrementing it.  For types like `int` it doesn't matter and the compiler is smart enough to optimize it anyway, but for classes / structs / complex types it can be faster.

Also if you compile without optimization flags it will probably be faster as well (I haven't tested).

Changing it in this context won't break anything and likely won't make anything faster, but I guess I was compelled to change it.  😛

Reply:
Fascinating!  Thanks for teaching me something new!